File: week11.kt
abstract class Enemy(val name: String, val level: Int, val maxHp: Int) {
    var currentHp: Int = maxHp

    abstract fun attack()
    abstract fun block()
    abstract fun move()
}

class Bandit(
    name: String, level: Int, maxHp: Int,
    private val gang: String, private val wealth: Int
) : Enemy(name, level, maxHp) {

    override fun attack() {
        println("$name from $gang attacks with ferocity!")
    }

    override fun block() {
        println("$name blocks with agility.")
    }

    override fun move() {
        println("$name moves to a new position stealthily.")
    }

    fun talk() {
        println("$name says, 'This is my turf!'")
    }
}

class Goblin(name: String, level: Int, maxHp: Int) : Enemy(name, level, maxHp) {
    override fun attack() {
        println("$name lunges forward and attacks!")
    }

    override fun block() {
        println("$name hides behind a rock to block.")
    }

    override fun move() {
        println("$name scuttles around quickly.")
    }

    fun climb() {
        println("$name climbs a tree with ease.")
    }

    fun eat() {
        println("$name chomps noisily on some food.")
    }

    fun screech() {
        println("$name lets out a piercing screech!")
    }
}

fun main() {
    val bandit = Bandit("Bandit King", 5, 100, "Black Scorpions", 500)
    bandit.attack()
    bandit.talk()

    val goblin = Goblin("Sneaky Goblin", 2, 50)
    goblin.move()
    goblin.screech()
    goblin.climb()
}

File: week11.rs
trait Enemy {
    fn attack(&self);
    fn block(&self);
    fn move_character(&self);
}

struct Bandit {
    name: String,
    level: u32,
    max_hp: u32,
    current_hp: u32,
    gang: String,
    wealth: u32,
}

impl Enemy for Bandit {
    fn attack(&self) {
        println!("{} from {} attacks with ferocity!", self.name, self.gang);
    }

    fn block(&self) {
        println!("{} blocks with agility.", self.name);
    }

    fn move_character(&self) {
        println!("{} moves to a new position stealthily.", self.name);
    }
}

impl Bandit {
    fn new(name: &str, level: u32, max_hp: u32, gang: &str, wealth: u32) -> Self {
        Bandit {
            name: name.to_string(),
            level,
            max_hp,
            current_hp: max_hp,
            gang: gang.to_string(),
            wealth,
        }
    }

    fn talk(&self) {
        println!("{} says, 'This is my turf!'", self.name);
    }
}

struct Goblin {
    name: String,
    level: u32,
    max_hp: u32,
    current_hp: u32,
}

impl Enemy for Goblin {
    fn attack(&self) {
        println!("{} lunges forward and attacks!", self.name);
    }

    fn block(&self) {
        println!("{} hides behind a rock to block.", self.name);
    }

    fn move_character(&self) {
        println!("{} scuttles around quickly.", self.name);
    }
}

impl Goblin {
    fn new(name: &str, level: u32, max_hp: u32) -> Self {
        Goblin {
            name: name.to_string(),
            level,
            max_hp,
            current_hp: max_hp,
        }
    }

    fn climb(&self) {
        println!("{} climbs a tree with ease.", self.name);
    }

    fn eat(&self) {
        println!("{} chomps noisily on some food.", self.name);
    }

    fn screech(&self) {
        println!("{} lets out a piercing screech!", self.name);
    }
}

fn main() {
    let bandit = Bandit::new("Bandit King", 5, 100, "Black Scorpions", 500);
    bandit.attack();
    bandit.talk();

    let goblin = Goblin::new("Sneaky Goblin", 2, 50);
    goblin.move_character();
    goblin.screech();
    goblin.climb();
}
